datastore-benchmarks/benchmarks/b3.js

```diff
@@ -0,0 +1,341 @@
+    const update = JSON.stringify(Automerge.getChanges(doc, updatedDoc))
+    updates.push(update)
+    docs[i] = updatedDoc
+  }
+  for (let i = 0; i < updates.length; i++) {
+    docs[0] = Automerge.applyChanges(docs[0], JSON.parse(updates[i]))
+  }
+  benchmarkTime('automerge', `${id} (time)`, () => {
+    for (let i = 0; i < updates.length; i++) {
+      docs[1] = Automerge.applyChanges(docs[1], JSON.parse(updates[i]))
+    }
+  })
+  check(docs.slice(0, 2))
+  setBenchmarkResult('automerge', `${id} (updateSize)`, `${updates.reduce((len, update) => len + update.length, 0)} bytes`)
+  const encodedState = Automerge.save(docs[0])
+  const documentSize = encodedState.length
+  setBenchmarkResult('automerge', `${id} (docSize)`, `${documentSize} bytes`)
+  benchmarkTime('automerge', `${id} (parseTime)`, () => {
+    Automerge.load(encodedState)
+  })
+}
+
+const benchmarkLumino = (id, changeDoc, check) => {
+  const stores = []
+  const adapters = []
+  const tables = []
+  const updates = []
+  let schema = {
+    id: 'test-schema',
+    fields: {
+      map: Fields.Map(),
+      array: Fields.List()
+    }
+  }
+  for (let i = 0; i < sqrtN; i++) {
+    const adapter = new InMemoryServerAdapter()
+    const store = Datastore.create({
+      id: i,
+      schemas: [schema],
+      adapter: adapter
+    });
+    store.changed.connect((_, change) => {
+      if (store.id === change.storeId) {
+        updates.push(adapter.transactions[change.transactionId])
+      }
+    })
+    adapters.push(adapter)
+    stores.push(store)
+    tables.push(store.get('test-schema'))
+  }
+  for (let i = 0; i < tables.length; i++) {
+    stores[i].beginTransaction()
+    changeDoc(tables[i], i)
+    stores[i].endTransaction()
+  }
+  // sync client 0 for reference
+  for (let i = 0; i < updates.length; i++) {
+    if (adapters[0].onRemoteTransaction) {
+      adapters[0].onRemoteTransaction(updates[i])
+    }
+  }
+  benchmarkTime('lumino', `${id} (time)`, () => {
+    for (let i = 0; i < updates.length; i++) {
+      for (let i = 0; i < updates.length; i++) {
+        if (adapters[1].onRemoteTransaction) {
+          adapters[1].onRemoteTransaction(updates[i])
+        }
+      }
+    }
+  })
+  t.assert(updates.length === sqrtN)
+  check(tables.slice(0, 2))
+  setBenchmarkResult('lumino', `${id} (updateSize)`, `${updates.reduce((len, update) => len + transactionSize(update), 0)} bytes`)
+  let adapter3 = new InMemoryServerAdapter()
+  let store3 = Datastore.create({
+    id: 10000000,
+    schemas: [schema],
+    adapter: adapter3
+  });
+  let documentSize = 0
+  for (let id in adapters[0].transactions) {
+    documentSize += transactionSize(adapters[0].transactions[id])
+  }
+  setBenchmarkResult('lumino', `${id} (docSize)`, `${documentSize} bytes`)
+  benchmarkTime('lumino', `${id} (parseTime)`, () => {
+    for (let i = 0; i < updates.length; i++) {
+      if (adapter3.onRemoteTransaction) {
+        adapter3.onRemoteTransaction(updates[i])
```

Comment:
Here, you apply the document updates without parsing them. This is a bit unfair because the other frameworks need to parse the document update here. 

Both Yjs and Automerge invested significant time to make encoding/parsing fast. Document updates are sent as binary blobs or as strings. JSON needs to be encoded first. If you would use JSON encoding here, your benchmark results would be 3 times slower.